kernel/src/process/program_loader/elf/load_elf.rs

```diff
@@ -133,118 +143,169 @@ fn init_and_map_vmos(
     } else {
-        0
+        // Not relocatable object. Map as-is.
+        elf_va_range.clone()
     };
+
+    let relocated_range =
+        RelocatedRange::new(elf_va_range, map_range.start).expect("Mapped range overflows");
+
     for program_header in &elf.program_headers {
-        let type_ = program_header
-            .get_type()
-            .map_err(|_| Error::with_message(Errno::ENOEXEC, "parse program header type fails"))?;
+        let type_ = program_header.get_type().map_err(|_| {
+            Error::with_message(Errno::ENOEXEC, "Failed to parse the program header")
+        })?;
         if type_ == program::Type::Load {
             check_segment_align(program_header)?;
-            map_segment_vmo(program_header, elf_file, root_vmar, base_addr)?;
+
+            let map_at = relocated_range
+                .relocated_addr_of(program_header.virtual_addr as Vaddr)
+                .expect("Address not covered by `get_range_for_all_segments`");
+
+            map_segment_vmo(program_header, elf_file, root_vmar, map_at)?;
         }
     }
-    Ok(base_addr)
+
+    Ok(relocated_range)
 }
 
-fn base_map_addr(elf: &Elf, root_vmar: &Vmar<Full>) -> Result<Vaddr> {
-    let elf_size = elf
-        .program_headers
-        .iter()
-        .filter_map(|program_header| {
-            if let Ok(type_) = program_header.get_type()
-                && type_ == program::Type::Load
-            {
-                let ph_max_addr = program_header.virtual_addr + program_header.mem_size;
-                Some(ph_max_addr as usize)
-            } else {
-                None
-            }
+/// A virtual range and its relocated address.
+pub struct RelocatedRange {
+    original_range: Range<Vaddr>,
+    relocated_start: Vaddr,
+}
+
+impl RelocatedRange {
+    /// Creates a new `RelocatedRange`.
+    ///
+    /// If the relocated address overflows, it will return `None`.
+    pub fn new(original_range: Range<Vaddr>, relocated_start: Vaddr) -> Option<Self> {
+        relocated_start.checked_add(original_range.len())?;
+        Some(Self {
+            original_range,
+            relocated_start,
         })
+    }
+
+    /// Gets the relocated address of an address in the original range.
+    ///
+    /// If the provided address is not in the original range, it will return `None`.
+    pub fn relocated_addr_of(&self, addr: Vaddr) -> Option<Vaddr> {
+        if self.original_range.contains(&addr) {
+            Some(addr - self.original_range.start + self.relocated_start)
+        } else {
+            None
+        }
+    }
+}
+
+/// Returns the range that covers all segments in the ELF file.
+///
+/// The range must be tight, i.e., will not include any padding bytes. So the
+/// boundaries may not be page-aligned.
+fn get_range_for_all_segments(elf: &Elf) -> Result<Range<Vaddr>> {
+    let loadable_ranges_iter = elf.program_headers.iter().filter_map(|ph| {
+        if let Ok(type_) = ph.get_type()
+            && type_ == program::Type::Load
```

Comment:
This condition can be simplified: 

`if let Ok(program::Type::Load) = ph.get_type()`